web_english/text/maping_text.py

```diff
@@ -0,0 +1,124 @@
+import json
+import re
+import urllib.request
+
+from pydub import AudioSegment
+
+from config import Config
+
+
+def chunk_audiofile(audiofile, title):
+    audio = AudioSegment.from_mp3(audiofile)
+    length_audio = len(audio)
+    counter = 1
+    interval = 3000
+    start = 0
+    end = 0
+    list_chunks = []
+    for i in range(0, length_audio, interval):
+        if i == 0:
+            start = 0
+            end = interval
+        else:
+            start = end
+            end = start + interval
+        if end >= length_audio:
+            end = length_audio
+        chunk = audio[start:end]
+        filename = f'chunks/{title}chunk{str(counter)}.ogg'
+        list_chunks.append(filename)
+        chunk.export(filename, format='ogg')
+        print(f"Processing {title}chunk{str(counter)}. Start = {str(start)} End = {str(end)}")
+        counter = counter + 1
+    return list_chunks
+
+
+def send_ya_speech_kit(*args):
+    list_chunks_result = []
+    for chunk in args:
+        with open(chunk, "rb") as f:
+            data = f.read()
+
+            params = "&".join([
+                "topic=general",
+                "folderId=%s" % Config.FOLDER_ID,
+                "lang=en-US"
+            ])
+
+            url = urllib.request.Request("https://stt.api.cloud.yandex.net/speech/v1/stt:recognize?%s" % params,
+                                         data=data)
+            url.add_header("Authorization", "Api-Key %s" % Config.API_KEY)
+
+            responseData = urllib.request.urlopen(url).read().decode('UTF-8')
+            decodedData = json.loads(responseData)
+
+            if decodedData.get("error_code") is None:
+                list_chunks_result.append(decodedData.get("result"))
+    return list_chunks_result
+
+
+# На вход принемаем оригинальный текст и список распознанных отрывков
+def maping_text(text, *args):
+    # Убираем из текста все знаки препинания и разбиваем по словам
+    split_text = re.sub("[.,!?;:]", "", text).lower().split()
+
+    # Та самая секунда, на которой находится диктор
+    second = 0
+
+    # Номер слова и само слово в тексте, на котором находится диктор
+    last_word = [0, None, second]
+    list_last_word = [last_word]
+
+    for recognized in args:
+        # Каждый отрывок - это 3 секунды чтения диктора
+        second += 3
+
+        # Отрезок оригинального текста, который будет сравниваться с
+        # определенным распознанным отрывком
+        cut_split_text = split_text[last_word[0]:last_word[0] + 15]
```

Comment:
15 - magic number https://ru.wikipedia.org/wiki/%D0%9C%D0%B0%D0%B3%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%BE%D0%B5_%D1%87%D0%B8%D1%81%D0%BB%D0%BE_(%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)